pkg/k8s/identitybackend/identity_test.go

```diff
@@ -82,3 +92,128 @@ func (s *K8sIdentityBackendSuite) TestSanitizeK8sLabels(c *C) {
+	return v2.CiliumIdentity{
+		ObjectMeta: v1.ObjectMeta{
+			Name: strconv.Itoa(id),
+		},
+		SecurityLabels: getLabelsMap(labels),
+	}
+}
+
+func TestGetIdentity(t *testing.T) {
+	simpleMap := map[string]string{"key": "value"}
+	simpleMap2 := map[string]string{"ke2": "value2"}
+	simpleMap3 := map[string]string{"key3": "value3"}
+	duplicateMap1 := map[string]string{"key": "foo=value"}
+	duplicateMap2 := map[string]string{"key=foo": "value"}
+
+	testCases := []struct {
+		desc         string
+		identities   []v2.CiliumIdentity
+		requestedKey allocator.AllocatorKey
+		expectedId   string
+	}{
+		{
+			desc:         "Simple case",
+			identities:   []v2.CiliumIdentity{createCiliumIdentity(10, simpleMap)},
+			requestedKey: getLabelsKey(simpleMap),
+			expectedId:   "10",
+		},
+		{
+			desc: "Multiple identities",
+			identities: []v2.CiliumIdentity{
+				createCiliumIdentity(10, simpleMap),
+				createCiliumIdentity(11, simpleMap2),
+				createCiliumIdentity(12, simpleMap3),
+			},
+			requestedKey: getLabelsKey(simpleMap2),
+			expectedId:   "11",
+		},
+		{
+			desc: "Duplicated identity",
+			identities: []v2.CiliumIdentity{
+				createCiliumIdentity(10, duplicateMap1),
+				createCiliumIdentity(11, duplicateMap1),
+			},
+			requestedKey: getLabelsKey(duplicateMap1),
+			expectedId:   "10",
+		},
+		{
+			desc: "Duplicated key",
+			identities: []v2.CiliumIdentity{
+				createCiliumIdentity(10, duplicateMap1),
+				createCiliumIdentity(11, duplicateMap2),
+			},
+			requestedKey: getLabelsKey(duplicateMap2),
+			expectedId:   "11",
+		},
+		{
+			desc:         "No identities",
+			identities:   []v2.CiliumIdentity{},
+			requestedKey: getLabelsKey(simpleMap),
+			expectedId:   idpool.NoID.String(),
+		},
+		{
+			desc: "Identity not found",
+			identities: []v2.CiliumIdentity{
+				createCiliumIdentity(10, simpleMap),
+				createCiliumIdentity(11, simpleMap2),
+			},
+			requestedKey: getLabelsKey(simpleMap3),
+			expectedId:   idpool.NoID.String(),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			_, client := k8sClient.NewFakeClientset()
+			backend, err := NewCRDBackend(CRDBackendConfiguration{
+				NodeName: "some-node",
+				Store:    nil,
+				Client:   client,
+				KeyFunc:  (&key.GlobalIdentity{}).PutKeyFromMap,
+			})
+			ctx := context.Background()
+			stopChan := make(chan struct{}, 1)
+			defer func() {
+				stopChan <- struct{}{}
+			}()
+			go backend.ListAndWatch(ctx, FakeHandler{}, stopChan)
+			if err != nil {
+				t.Fatalf("Can't create CRD Backedn: %s", err)
+			}
+
+			for _, identity := range tc.identities {
+				_, err = client.CiliumV2().CiliumIdentities().Create(ctx, &identity, v1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("Can't create identity %s: %s", identity.Name, err)
+				}
+			}
+			// Wait for watcher to process the identities in the background
+			time.Sleep(100 * time.Millisecond)
+
+			id, err := backend.Get(ctx, tc.requestedKey)
```

Comment:
Random sleep durations can lead to test flakes. Have you instead considered adding retries with bounded time?

Reply:
Good idea, fixed.